File: ts/src/standardLibrary/standardAssertLibrary.ts
import { Scope, IReadOnlyScope } from "../scope";
import { BuiltinFunctionValue } from "../values/builtinFunctionValue";
import { ObjectValue, ObjectValueMap } from "../values/objectValue";

export const assertScope: IReadOnlyScope = createAssertScope();

export function createAssertScope()
{
    const result: Scope = new Scope();

    const assertFunctions: ObjectValueMap =
    {
        true: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getBool(0);
            if (!top)
            {
                vm.running = false;
                console.error(vm.createStackTrace().join('\n'));
                console.error('Assert expected true');
            }
        }, 'assert.true'),

        false: new BuiltinFunctionValue((vm, args) =>
        {
            const top = args.getBool(0);
            if (top)
            {
                vm.running = false;
                console.error(vm.createStackTrace().join('\n'));
                console.error('Assert expected false');
            }
        }, 'assert.false'),

        equals: new BuiltinFunctionValue((vm, args) =>
        {
            const expected = args.getIndex(0);
            const actual = args.getIndex(1);
            if (expected.compareTo(actual) !== 0)
            {
                vm.running = false;
                console.error(vm.createStackTrace().join('\n'));
                console.error(`Assert expected equals:\nExpected: ${expected.toString()}\nActual: ${actual.toString()}`);
            }
        }, 'assert.equals'),

        notEquals: new BuiltinFunctionValue((vm, args) =>
        {
            const expected = args.getIndex(0);
            const actual = args.getIndex(1);
            if (expected.compareTo(actual) === 0)
            {
                vm.running = false;
                console.error(vm.createStackTrace().join('\n'));
                console.error(`Assert expected not equals:\nActual: ${actual.toString()}`);
            }
        }, 'assert.notEquals')
    };

    result.tryDefine('assert', new ObjectValue(assertFunctions));

    return result;
}

File: dotnet/VM/StandardLibrary/StandardAssertLibrary.cs
using System;
using System.Collections.Generic;

namespace LysitheaVM
{
    public static class StandardAssertLibrary
    {
        #region Fields
        public static readonly IReadOnlyScope Scope = CreateScope();
        #endregion

        #region Methods
        public static Scope CreateScope()
        {
            var result = new Scope();

            var assertFunctions = new Dictionary<string, IValue>
            {
                {"true", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<BoolValue>(0);
                    if (!top.Value)
                    {
                        vm.Running = false;
                        Console.WriteLine(string.Join("\n", vm.CreateStackTrace()));
                        Console.WriteLine($"Assert expected true");
                    }
                }, "assert.true")},

                {"false", new BuiltinFunctionValue((vm, args) =>
                {
                    var top = args.GetIndex<BoolValue>(0);
                    if (top.Value)
                    {
                        vm.Running = false;
                        Console.WriteLine(string.Join("\n", vm.CreateStackTrace()));
                        Console.WriteLine($"Assert expected false");
                    }
                }, "assert.false")},

                {"equals", new BuiltinFunctionValue((vm, args) =>
                {
                    var expected = args.GetIndex(0);
                    var actual = args.GetIndex(1);
                    if (expected.CompareTo(actual) != 0)
                    {
                        vm.Running = false;
                        Console.WriteLine(string.Join("\n", vm.CreateStackTrace()));
                        Console.WriteLine($"Assert expected equals:\nExpected: {expected.ToString()}\nActual: {actual.ToString()}");
                    }
                }, "assert.equals")},

                {"notEquals", new BuiltinFunctionValue((vm, args) =>
                {
                    var expected = args.GetIndex(0);
                    var actual = args.GetIndex(1);
                    if (expected.CompareTo(actual) == 0)
                    {
                        vm.Running = false;
                        Console.WriteLine(string.Join("\n", vm.CreateStackTrace()));
                        Console.WriteLine($"Assert expected not equals:\nActual: {actual.ToString()}");
                    }
                }, "assert.notEquals")}
            };

            result.TryDefine("assert", new ObjectValue(assertFunctions));

            return result;
        }
        #endregion
    }
}